op-batcher/batcher/channel_manager.go

```diff
@@ -133,23 +126,54 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
 	s.log.Debug("marked transaction as confirmed", "id", id, "block", inclusionBlock)
 }
 
-// removePendingChannel removes the given completed channel from the manager's state.
-func (s *channelManager) removePendingChannel(channel *channel) {
-	if s.currentChannel == channel {
-		s.currentChannel = nil
+// rewindToBlock updates the blockCursor to point at
+// the block with the supplied hash, only if that block exists
+// in the block queue and the blockCursor is ahead of it.
+// Panics if the block is not in state.
+func (s *channelManager) rewindToBlock(block eth.BlockID) {
+	idx := block.Number - s.blocks[0].Number().Uint64()
+	if s.blocks[idx].Hash() == block.Hash && idx < uint64(s.blockCursor) {
+		s.blockCursor = int(idx)
+	} else {
+		panic("tried to rewind to nonexistent block")
 	}
-	index := -1
-	for i, c := range s.channelQueue {
-		if c == channel {
-			index = i
+}
+
+// handleChannelTimeout rewinds the channelManager's blockCursor
+// to point at the first block added to the provided channel,
+// and removes any channels which are newer than the provided channel.
+func (s *channelManager) handleChannelTimeout(c *channel) {
+	if len(c.channelBuilder.blocks) > 0 {
+		// This is usually true, but there is an edge case
+		// where a channel timed out before any blocks got added.
+		// In that case we end up with an empty frame (header only),
+		// and there are no blocks to requeue.
+		blockID := eth.ToBlockID(c.channelBuilder.blocks[0])
+		for _, block := range c.channelBuilder.blocks {
+			s.metr.RecordL2BlockInPendingQueue(block)
+		}
+		s.rewindToBlock(blockID)
+	} else {
+		s.log.Debug("channelManager.handleChannelTimeout: channel had no blocks")
+	}
+
+	// Trim back to and including the provided channel.
+	for i := range s.channelQueue {
+		if s.channelQueue[i] == c {
+			if len(s.channelQueue) > i {
+				// keep older channels if there are any,
+				// in case we later learn they also timed out.
+				s.channelQueue = s.channelQueue[i+1:]
+			} else {
+				// if this is the oldest channel, clear the queue
+				s.channelQueue = s.channelQueue[:0]
+			}
```

Comment:
So maybe this becomes
```suggestion
			s.channelQueue = s.channelQueue[:i+1]
```
?

Reply:
Yes https://github.com/ethereum-optimism/optimism/pull/12390/commits/74ed8b81609a1bc7b1c143175765d6394c7056f6

Reply:
Although note we need to exclude the actual channel which got invalidated https://github.com/ethereum-optimism/optimism/pull/12390/commits/7b17b1c693de904f2ec1cf36317db39348841e87.